shared/teams/team/settings/index.js

```diff
@@ -0,0 +1,272 @@
+// @flow
+import * as React from 'react'
+import * as Types from '../../../constants/types/teams'
+import {Box, Button, Checkbox, ScrollView, Text} from '../../../common-adapters'
+import {globalColors, globalMargins, globalStyles} from '../../../styles'
+import {isMobile} from '../../../constants/platform'
+
+// initial settings
+type Props = {
+  ignoreAccessRequests: boolean,
+  publicityAnyMember: boolean,
+  publicityMember: boolean,
+  publicityTeam: boolean,
+  openTeam: boolean,
+  openTeamRole: Types.TeamRoleType,
+  savePublicity: any => void,
+  setOpenTeamRole: (
+    newOpenTeamRole: Types.TeamRoleType,
+    setNewOpenTeamRole: (Types.TeamRoleType) => void
+  ) => void,
+  yourOperations: Types.TeamOperations,
+  waitingForSavePublicity: boolean,
+}
+
+type NewSettings = {
+  newIgnoreAccessRequests: boolean,
+  newPublicityAnyMember: boolean,
+  newPublicityMember: boolean,
+  newPublicityTeam: boolean,
+  newOpenTeam: boolean,
+  newOpenTeamRole: Types.TeamRoleType,
+}
+
+// new settings
+type State = NewSettings & {
+  publicitySettingsChanged: boolean,
+}
+
+export class Settings extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props)
+    this.state = {
+      newIgnoreAccessRequests: this.props.ignoreAccessRequests,
+      newPublicityAnyMember: this.props.publicityAnyMember,
+      newPublicityMember: this.props.publicityMember,
+      newPublicityTeam: this.props.publicityTeam,
+      newOpenTeam: this.props.openTeam,
+      newOpenTeamRole: this.props.openTeamRole,
+      publicitySettingsChanged: false,
+    }
+  }
+
+  componentWillReceiveProps(nextProps: Props) {
+    // We just got new settings for this team, reset any user selections
+    // to reflect the actual settings.
+    this.setState({
+      newIgnoreAccessRequests: nextProps.ignoreAccessRequests,
+      newPublicityAnyMember: nextProps.publicityAnyMember,
+      newPublicityMember: nextProps.publicityMember,
+      newPublicityTeam: nextProps.publicityTeam,
+      newOpenTeam: nextProps.openTeam,
+      newOpenTeamRole: nextProps.openTeamRole,
+      publicitySettingsChanged: false,
+    })
+  }
+
+  onSavePublicity = () => {
+    this.props.savePublicity({
+      ignoreAccessRequests: this.state.newIgnoreAccessRequests,
+      openTeam: this.state.newOpenTeam,
+      openTeamRole: this.state.newOpenTeamRole,
+      publicityAnyMember: this.state.newPublicityAnyMember,
+      publicityMember: this.state.newPublicityMember,
+      publicityTeam: this.state.newPublicityTeam,
+    })
+  }
+
+  setSettings = (key: $Keys<State>) => (newSetting: *) => {
+    // $FlowIssue with TeamRoleType not being a boolean
```

Comment:
I think it's better to split this up into two functions, one for bools and one for `newOpenTeamRole`. The annoying thing is that you might have to list the union type explicitly for the arguments, but I think that's better than a FlowIssue.

Or you might be able to do something fancy with templates and `$PropertyType`...

Reply:
I tried `$PropertyType<T, k>` but it only accepts a string literal for `k` 🤔. I changed team role to call setState directly instead